src/transformers/NestedLoopTransformation.cc

```diff
@@ -0,0 +1,213 @@
+    }();
+
+    for (std::size_t i = 1; i < errorPath.size(); ++i) {
+        SymRef target = graph.getTarget(errorPath[i].clauseId);
+        auto loopNode = std::find_if(loops.begin(), loops.end(),
+                                     [&target](const WitnessInfo & x) { return x.loopVertex == target; });
+        if (loopNode != loops.end()) {
+            auto it = std::find_if(allVertices.begin(), allVertices.end(), [&](auto vertex) {
+                if (loopNode->locations.find(vertex) == loopNode->locations.end()) { return false; }
+                auto varName = ".loc_" + std::to_string(vertex.x);
+                auto vertexVar = logic.mkBoolVar(varName.c_str());
+                vertexVar = tm.getVarVersionZero(vertexVar);
+                vertexVar = tm.sendVarThroughTime(vertexVar, i + 1);
+                return model->evaluate(vertexVar) == logic.getTerm_true();
+            });
+            assert(it != allVertices.end());
+            pathVertices.push_back(*it);
+        } else {
+            pathVertices.push_back(initialGraph.getTarget(errorPath[i].clauseId));
+        };
+    }
+
+    // Build error path from the vertices
+    std::vector<EId> errorEdges;
+    auto adj = AdjacencyListsGraphRepresentation::from(initialGraph);
+    for (std::size_t i = 1; i < pathVertices.size(); ++i) {
+        auto source = pathVertices[i - 1];
+        auto target = pathVertices[i];
+        auto const & outgoing = adj.getOutgoingEdgesFor(source);
+        auto it = std::find_if(outgoing.begin(), outgoing.end(),
+                               [&](EId eid) { return target == initialGraph.getTarget(eid); });
+        assert(it != outgoing.end());
+        errorEdges.push_back(*it);
+        // TODO: deal with multiedges properly
+        if (std::find_if(it + 1, outgoing.end(), [&](EId eid) { return target == initialGraph.getTarget(eid); }) !=
+            outgoing.end()) {
+            // Bail out in this case
+            return NoWitness{"Could not backtranslate invalidity witness in single-loop transformation"};
+        }
+    }
+    ErrorPath ep;
+    ep.setPath(std::move(errorEdges));
+    return InvalidityWitness::fromErrorPath(ep, initialGraph);
+}
+
+NestedLoopTransformation::WitnessBackTranslator::ErrorOr<ValidityWitness>
+NestedLoopTransformation::WitnessBackTranslator::translateInvariant(ValidityWitness wtns) {
+    Logic & logic = graph.getLogic();
+    TimeMachine timeMachine(logic);
+    auto vertices = graph.getVertices();
+    auto oldVertices = initialGraph.getVertices();
+    TermUtils utils(logic);
+    TermUtils::substitutions_map substitutions;
+    for (auto vertex : oldVertices) {
+        if (vertex == graph.getEntry()) { continue; }
+        for (auto loop : loops) {
+            try {
+                PTRef locationVar = loop.locations.at(vertex);
+                substitutions.insert({timeMachine.getUnversioned(locationVar), logic.getTerm_false()});
+            } catch (const std::out_of_range & ex) { continue; }
```

Comment:
Please don't use exceptions to check if some element is in a map.

Reply:
Fixed